src/runtime/program/precompile_programs/lib.zig

```diff
@@ -0,0 +1,175 @@
+    feature_set: sig.runtime.FeatureSet,
+) (PrecompileProgramError || error{OutOfMemory})!void {
+    // could remove this alloc by passing in the transaction in directly, but maybe less clean
+    var instruction_datas: ?[]const []const u8 = null;
+    defer if (instruction_datas) |instr_datas| allocator.free(instr_datas);
+
+    for (transaction.msg.instructions) |instruction| {
+        const program_id = transaction.msg.account_keys[instruction.program_index];
+        for (PRECOMPILES) |precompile| {
+            if (!precompile.program_id.equals(&program_id)) continue;
+
+            const precompile_feature_enabled = precompile.required_feature == null or
+                feature_set.active.contains(precompile.required_feature.?);
+            if (!precompile_feature_enabled) continue;
+
+            const datas = instruction_datas orelse blk: {
+                const buf = try allocator.alloc([]const u8, transaction.msg.instructions.len);
+                for (transaction.msg.instructions, 0..) |instr, i| buf[i] = instr.data;
+                instruction_datas = buf;
+                break :blk buf;
+            };
+
+            try precompile.function(instruction.data, datas);
+        }
+    }
+}
+
+pub const PrecompileFn = fn (
+    current_instruction_data: []const u8,
+    all_instruction_datas: []const []const u8,
+) PrecompileProgramError!void;
+
+pub const Precompile = struct {
+    program_id: Pubkey,
+    function: *const PrecompileFn,
+    required_feature: ?Pubkey,
+};
+
+// custom errors
+// https://github.com/anza-xyz/agave/blob/a8aef04122068ec36a7af0721e36ee58efa0bef2/sdk/precompile-error/src/lib.rs#L6
+pub const PrecompileProgramError = error{
+    InvalidPublicKey,
+    InvalidRecoveryId,
+    InvalidSignature,
+    InvalidDataOffsets,
+    InvalidInstructionDataSize,
+};
+
+pub fn getInstructionValue(
+    T: type,
+    current_instruction_data: []const u8,
+    all_instruction_datas: []const []const u8,
+    instruction_idx: u16,
+    offset: usize,
+) error{InvalidDataOffsets}!*align(1) const T {
+    return @ptrCast(try getInstructionData(
+        @sizeOf(T),
+        current_instruction_data,
+        all_instruction_datas,
+        instruction_idx,
+        offset,
+    ));
+}
+
+// https://github.com/firedancer-io/firedancer/blob/af74882ffb2c24783a82718dbc5111a94e1b5f6f/src/flamenco/runtime/program/fd_precompiles.c#L74
+pub fn getInstructionData(
+    len: usize,
+    current_instruction_data: []const u8,
+    all_instruction_datas: []const []const u8,
+    instruction_idx: u16,
+    offset: usize,
+) error{InvalidDataOffsets}![]const u8 {
+    const data: []const u8 = if (instruction_idx == std.math.maxInt(u16))
+        current_instruction_data
+    else data: {
+        if (instruction_idx >= all_instruction_datas.len) return error.InvalidDataOffsets;
+        break :data all_instruction_datas[instruction_idx];
+    };
+
+    if (offset + len > data.len) return error.InvalidDataOffsets;
```

Comment:
probs want saturating add here

Reply:
Changed to saturating 41dbdff2327181564202bac43005ba2cf3959cfd